k8s-releng-prod: Signer accounts, access and IAM API

puerco opened the pull request:
Following our discussions on the RelEng meetings and in https://github.com/kubernetes/k8s.io/issues/3461#issuecomment-1056017938 this PR sets up some cloud resources for the signing MVP on `k8s-releng-prod`. 

Namely does the following:

1. Creates the staging and production signer accounts
2. Ensures the IAM Service Account Credentials API is enabled
4. Grants access to the signer accounts to the following:
  - Grants the release process service account access to the stage signer acct
  - Grants the image promoter service accounts (file+image) access to the production signer 
5. Ensures the cross-project impersonation constraint is not enforced in the project

Part of https://github.com/kubernetes/release/issues/2445

/assign @ameukam 
/cc @kubernetes/sig-release-leads @kubernetes/release-engineering 
/sig k8s-infra
/sig release
/priority important-soon
/area bash
/area infra


Signed-off-by: Adolfo García Veytia (Puerco) <adolfo.garcia@uservers.net>

ameukam commented on infra/gcp/bash/ensure-prod-storage.sh:
```diff
@@ -55,6 +55,12 @@ GCR_AUDIT_TEST_PROD_PROJECT=$(k8s_infra_project "prod" "k8s-gcr-audit-test-prod"
 
+# The staging signer service account. To sign artifacts during staging
+STAGING_SIGNER_SVCACCT="k8s-releng"
+RELEASE_PROCESS_PROJECT="kubernetes-release-test"
```
Do we have a reason to use a Google project instead of a community-owned project ? we recently switch for to `k8s-staging-releng-test` https://github.com/kubernetes/release/pull/2414 ?

puerco replied:
Dropped the google owned project.

puerco commented:
I've repurposed the PR to create the required resources in `k8s-releng-prod` as discussed. PTAL @ameukam @justaugustus

puerco commented:
@ameukam can we move forward with this one to unblock signatures during staging?

Ref: https://github.com/kubernetes/release/pull/2397#issuecomment-1058579002

ameukam commented on infra/gcp/bash/ensure-releng.sh:
```diff
@@ -44,6 +44,103 @@ if [ $# = 0 ]; then
+    promoter_file_account="$(svc_acct_email "${PROMOTER_PROJECT}" "${FILE_PROMOTER_SVCACCT}")"
+    promoter_image_account="$(svc_acct_email "${PROMOTER_PROJECT}" "${IMAGE_PROMOTER_SVCACCT}")"
+    
+    ensure_serviceaccount_role_binding \
+        "${prod_sign_account}" \
+        "serviceAccount:${promoter_file_account}" \
+        "roles/iam.serviceAccountTokenCreator"
+    
+    ensure_serviceaccount_role_binding \
+        "${prod_sign_account}" \
+        "serviceAccount:${promoter_image_account}" \
+        "roles/iam.serviceAccountTokenCreator"
+}
```
Looks...like a duplicate ?

puerco replied:
Nope, files and images are promoted by different identities:

`k8s-infra-promoter` vs `k8s-infra-gcr-promoter`

ameukam commented on infra/gcp/bash/lib.sh:
```diff
@@ -80,10 +80,10 @@ readonly AUDITOR_INVOKER_SVCACCT="k8s-infra-gcr-auditor-invoker"
 
 # The service account name for the file promoter.
-readonly FILE_PROMOTER_SVCACCT="k8s-infra-promoter"
+export FILE_PROMOTER_SVCACCT="k8s-infra-promoter"
```
Why this is needed ?

puerco replied:
These variables are now used outside the script, in `ensure-releng.sh`

puerco replied:
Mmh I thought it was needed to access the vars outside the lib. Dropped the commit.

ameukam commented on infra/gcp/bash/ensure-releng.sh:
```diff
@@ -44,6 +44,103 @@ if [ $# = 0 ]; then
+    # Check if the constraint is enabled
+    local status_file
+    status_file=$(mktemp)
+    if gcloud --project="${project}" org-policies describe \
```
This action requires we enable the `orgpolicy.googleapis.com` for this project. I don't think we want to do this now.

puerco replied:
If we don't switch this constraint off, the promoter and the release process cannot impersonate the signing service accounts. This means that they cannot generate the OIDC tokens they need and cannot sign any artifacts.

puerco commented on infra/gcp/bash/ensure-releng.sh:
```diff
@@ -44,6 +44,103 @@ if [ $# = 0 ]; then
 
+RELEASE_PROCESS_CLOUDBUILD_SVCACCT="648026197307@cloudbuild.gserviceaccount.com"
+STAGING_SIGNER_SVCACCT="krel-staging"
+K8s_ORG_SIGNER_SVCACCT="trust"
```
This account does not have any permissions. It is only the identity provider.

In the future, when you verify an image or binary, the transparency log will show this identity as the signer.

This is the organization wide identity, it is not krel specific.

puerco commented on infra/gcp/bash/ensure-releng.sh:
```diff
@@ -44,6 +44,103 @@ if [ $# = 0 ]; then
+        echo "  rules:"
+        echo "  - enforce: false" 
+    } > "${policy_file}"
+    gcloud org-policies set-policy "${policy_file}" || return 1
```
Thanks! Done.

puerco commented:
/test pull-k8sio-verify

puerco commented:
@ameukam I've addressed your comments, please PTAL :)
(CI is broken for external reasons which I've fixed in https://github.com/kubernetes/k8s.io/pull/3495 )

puerco commented:
/test pull-k8sio-verify

ameukam commented:
/lgtm
/approve
